pkg/plugins/plugins.go

```diff
@@ -150,6 +151,9 @@ type JSONData struct {
 
 	// Backend (Datasource + Renderer + SecretsManager)
 	Executable string `json:"executable,omitempty"`
+
+	// Oauth App Service Registration
+	OauthServiceRegistration *oauth.ExternalServiceRegistration `json:"oauthServiceRegistration,omitempty"`
```

Comment:
What do you think here too @mgyongyosi? What's the minimal set of configuration that plugins need?

> Also related, this should go into https://github.com/grafana/grafana/blob/main/pkg/plugins/plugindef/plugindef.cue which means all code for plugin.json will be generated which falls in line with suggestion above to keep plugin.json contracts in this package.

Yep, that was part of the original PR but decided not to add that here to avoid more changes to review. I will open a follow-up PR once this is complete.

Reply:
> Specifying self or impersonate requires the enabled property to be set to true - to me should be enough with having self or impersonate set for them to be enabled.

We wanted to be explicit about the setting that's why we added the `enabled` property.

> What's the minimal set of configuration that plugins need?

One of those should be enabled and at least one permission has to be specified because without any permission the plugin cannot call Grafana.

> We need to carefully consider how we expose this in plugin.json. Maybe we should use another struct defined in this package here as the contract to allow a more user friendly/plugin specific config and later translate that into whatever we need. Similar to the Roles property above.

Let me spend some time on this. :) I'll get back to you.

Reply:
@marefr @mgyongyosi please take a look to the simplified version I just sent!

Reply:
> We wanted to be explicit about the setting that's why we added the enabled property.

Can you elaborate on this?

> One of those should be enabled and at least one permission has to be specified because without any permission the plugin cannot call Grafana.

This should still be validated within `SavePluginExternalService` or by plugin management before calling `SavePluginExternalService`. As a plugin dev would make sense Grafana refuses to load a plugin if this validation fails.

Reply:
Currently it's expected `externalServiceRegistration` property to be used in plugin.json, [example](https://github.com/grafana/grafana-plugin-examples/pull/138/files) is referring to `oauthServiceRegistration`. Which one do you prefer? Not so sure it needs to be specific to/mentioning OAuth terms? The more I look at this, I would maybe suggest simplifying all this even more and shrink the structure to make it easier to read/edit it, something like this:

```json
{
  "<externalServiceRegistration|oauthServiceRegistration|grafanaAuthentication|oauth|grafanaClient>": {
    "permissions": [
      { "action": "dashboards:create", "scope": "folders:uid:general" }
    ],
    "impersonation": {
      "group": true,
      "permissions": [
        { "action": "dashboards:create", "scope": "folders:*" },
        { "action": "dashboards:read", "scope": "dashboards:*" },
        { "action": "dashboards:read", "scope": "folders:*" },
        { "action": "folders:read", "scope": "folders:*" }
      ]
    }
  }
```

Reply:
sorry, I forgot to push the changes in the example to adapt to the new model I propose here. This is how it looks now:

```json
  "externalServiceRegistration": {
    "selfPermissions": [{ "action": "dashboards:create", "scope": "folders:uid:general" }],
    "impersonationPermissions": [
      { "action": "dashboards:create", "scope": "folders:*" },
      { "action": "dashboards:read", "scope": "dashboards:*" },
      { "action": "dashboards:read", "scope": "folders:*" },
      { "action": "folders:read", "scope": "folders:*" }
    ]
  }
```

the question is if this is too simplistic, is `group: true` needed?

Reply:
>> We wanted to be explicit about the setting that's why we added the enabled property.

>Can you elaborate on this?

I commented on this statement: 'Specifying self or impersonate requires the enabled property to be set to true - to me should be enough with having self or impersonate set for them to be enabled."
The reason we added the `enabled` property to `self` and `impersonation` was to be be explicit that the `self` and/or `impersonation` functionality is enabled.

> is group: true needed?

The `group` property specifies whether the plugin can list the impersonated users' teams.

> This should still be validated within SavePluginExternalService or by plugin management before calling SavePluginExternalService. As a plugin dev would make sense Grafana refuses to load a plugin if this validation fails.

Agree

Reply:
> The reason we added the enabled property to self and impersonation was to be be explicit that the self and/or impersonation functionality is enabled.

@mgyongyosi was this specifically thought of for plugins/plugin authors or for the `SaveExternalService` contract?